demux/demux_lavf.c

```diff
@@ -603,48 +603,50 @@ static void select_tracks(struct demuxer *demuxer, int start)
 static void export_replaygain(demuxer_t *demuxer, struct sh_stream *sh,
                               AVStream *st)
 {
-    AVPacketSideData *side_data = st->codecpar->coded_side_data;
-    int nb_side_data = st->codecpar->nb_coded_side_data;
-    for (int i = 0; i < nb_side_data; i++) {
-        AVReplayGain *av_rgain;
-        struct replaygain_data *rgain;
-        AVPacketSideData *src_sd = &side_data[i];
-
-        if (src_sd->type != AV_PKT_DATA_REPLAYGAIN)
-            continue;
+    struct replaygain_data *rgain = NULL;
+    const AVPacketSideData *side_data =
+        av_packet_side_data_get(st->codecpar->coded_side_data,
+                                st->codecpar->nb_coded_side_data,
+                                AV_PKT_DATA_REPLAYGAIN);
+    if (!side_data || !side_data->data)
+        return;
 
-        av_rgain = (AVReplayGain*)src_sd->data;
-        rgain    = talloc_ptrtype(demuxer, rgain);
-        rgain->track_gain = rgain->album_gain = 0;
-        rgain->track_peak = rgain->album_peak = 1;
-
-        // Set values in *rgain, using track gain as a fallback for album gain
-        // if the latter is not present. This behavior matches that in
-        // demux/demux.c's decode_rgain; if you change this, please make
-        // equivalent changes there too.
-        if (av_rgain->track_gain != INT32_MIN && av_rgain->track_peak != 0.0) {
-            // Track gain is defined.
-            rgain->track_gain = av_rgain->track_gain / 100000.0f;
-            rgain->track_peak = av_rgain->track_peak / 100000.0f;
-
-            if (av_rgain->album_gain != INT32_MIN &&
-                av_rgain->album_peak != 0.0)
-            {
-                // Album gain is also defined.
-                rgain->album_gain = av_rgain->album_gain / 100000.0f;
-                rgain->album_peak = av_rgain->album_peak / 100000.0f;
-            } else {
-                // Album gain is undefined; fall back to track gain.
-                rgain->album_gain = rgain->track_gain;
-                rgain->album_peak = rgain->track_peak;
-            }
-        }
+    const AVReplayGain *av_rgain =
+        (const AVReplayGain*)side_data->data;
+
+    bool track_data_available =
+        (av_rgain->track_gain != INT32_MIN && av_rgain->track_peak != 0);
+    bool album_data_available =
+        (av_rgain->album_gain != INT32_MIN && av_rgain->album_peak != 0);
```

Comment:
Is it possible to have gain without peak value?

Reply:
This is just the same check being moved from inside the ifs to variables, so unfortunately I don't have the answer to that.

For better or worse the behavior check-wise should at least be the same.